rust/kernel/types.rs

```diff
@@ -200,29 +206,66 @@ impl<T: PointerWrapper + Deref> PointerWrapper for Pin<T> {
 ///     pr_info!("example2 no early return\n");
 /// }
 /// ```
-pub struct ScopeGuard<T: FnOnce()> {
-    cleanup_func: Option<T>,
+///
+/// In the example below, we need a mutable object (the vector) to be accessible within the log
+/// function, so we wrap it in the [`ScopeGuard`]:
+/// ```
+/// # use kernel::prelude::*;
+/// # use kernel::ScopeGuard;
+/// fn example3(arg: bool) -> Result {
+///     let mut vec = ScopeGuard::new(Vec::new(), |v| pr_info!("vec had {} elements\n", v.len()));
+///
+///     vec.try_push(10u8)?;
+///     if arg {
+///         return Ok(());
+///     }
+///     vec.try_push(20u8)?;
+///     Ok(())
+/// }
+/// ```
+pub struct ScopeGuard<T, F: FnOnce(&mut T)> {
+    cleanup_func: Option<F>,
+    data: T,
```

Comment:
Maybe wrap this in `ManuallyDrop`? This would allow implementing `dismiss` a bit more safely by first taking the cleanup func to prevent the drop impl from accessing `data` and then reading the `data`. It would also avoid a memory leak when the cleanup func captures heap allocated data.